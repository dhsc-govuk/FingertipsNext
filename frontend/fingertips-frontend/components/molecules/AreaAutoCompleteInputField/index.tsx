import { getSearchSuggestions } from '@/components/forms/SearchForm/searchActions';
import { AreaDocument } from '@/lib/search/searchTypes';
import React, { useEffect, useState, useRef } from 'react';
import { AreaSearchInputField } from '@/components/molecules/AreaSearchInputField';
import { AreaAutoCompleteSuggestionPanel } from '@/components/molecules/AreaSuggestionPanel';
import styled from 'styled-components';
import { SearchParams, SearchStateParams } from '@/lib/searchStateManager';
import { AreaWithRelations } from '@/generated-sources/ft-api-client';

const MIN_SEARCH_SIZE = 3;
const DEBOUNCE_SEARCH_DELAY = 300;
const MAX_SUGGESTED_AREA_LIMIT = 20;

const StyleAreaAutoCompleteInputField = styled('div')({
  marginBottom: '1em',
});

interface AreaAutoCompleteInputFieldProps {
  inputFieldErrorStatus: boolean;
  searchState?: SearchStateParams;
  firstSelectedArea?: AreaWithRelations;
}

export function AreaAutoCompleteInputField({
  searchState,
  inputFieldErrorStatus = false,
  firstSelectedArea,
}: Readonly<AreaAutoCompleteInputFieldProps>) {
  const selectedAreasParams = searchState?.[SearchParams.AreasSelected];

  const [criteria, setCriteria] = useState<string>();
  const [searchAreas, setSearchAreas] = useState<AreaDocument[]>([]);

  const timeoutRef = useRef<number | null>(null);

  useEffect(() => {
    if (firstSelectedArea) {
      setCriteria(firstSelectedArea?.name ?? '');
    }

    const fetchSearchArea = async (criteria: string) => {
      if (criteria) {
        const areas = await getSearchSuggestions(criteria);
        setSearchAreas(areas.slice(0, MAX_SUGGESTED_AREA_LIMIT));
      }
    };

    const fetchCleanUp = () => {
      if (timeoutRef.current != null) {
        clearTimeout(timeoutRef.current);
        timeoutRef.current = null;
      }
    };
    fetchCleanUp();

    timeoutRef.current = setTimeout(
      (criteria: string) => {
        if (
          firstSelectedArea ||
          criteria == null ||
          criteria.length < MIN_SEARCH_SIZE
        ) {
          setSearchAreas([]);
          return;
        }
        fetchSearchArea(criteria);
      },
      DEBOUNCE_SEARCH_DELAY,
      criteria
    );

    return fetchCleanUp;
  }, [criteria, firstSelectedArea]);

  return (
    <StyleAreaAutoCompleteInputField>
      <AreaSearchInputField
        value={criteria}
        onTextChange={setCriteria}
        disabled={
          (selectedAreasParams && selectedAreasParams?.length > 0) ||
          searchState?.[SearchParams.GroupAreaSelected] === 'ALL'
        }
        hasError={inputFieldErrorStatus}
      />
      <AreaAutoCompleteSuggestionPanel
        suggestedAreas={searchAreas}
        searchHint={criteria ?? ''}
        searchState={searchState}
      />
    </StyleAreaAutoCompleteInputField>
  );
}
