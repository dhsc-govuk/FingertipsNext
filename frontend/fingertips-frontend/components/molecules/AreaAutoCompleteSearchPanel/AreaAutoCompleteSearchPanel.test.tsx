import React from 'react';
import { render } from '@testing-library/react';
import AreaAutoCompleteInputField from './index';

jest.mock('@/components/forms/SearchForm/searchActions', () => ({
  getSearchSuggestions: jest.fn(),
}));

describe('test AreaAutoCompleteSearchPanel', () => {
  const mockOnAreaSelected = jest.fn();
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('take a component snapshot', async () => {
    const { asFragment } = render(
      <AreaAutoCompleteInputField
        onAreaSelected={mockOnAreaSelected}
        inputFieldErrorStatus={false}
        defaultValue="Willington Medical Group"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
