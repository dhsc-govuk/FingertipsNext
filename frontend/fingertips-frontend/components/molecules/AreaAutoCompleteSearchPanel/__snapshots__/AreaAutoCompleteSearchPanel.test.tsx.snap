// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AreaAutoCompleteInputField take a component snapshot 1`] = `
<DocumentFragment>
  .c4 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  display: block;
  margin-bottom: 15px;
  color: #6f777b;
}

.c3 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  color: #0b0c0c;
  display: block;
  clear: none;
  padding-bottom: 2px;
  margin-bottom: 0;
}

.c5 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  box-sizing: border-box;
  width: 100%;
  height: 40px;
  margin-top: 0;
  padding: 5px;
  border: 2px solid #0b0c0c;
  border-radius: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin-bottom: 0;
}

.c5:focus {
  outline: 3px solid #ffdd00;
  outline-offset: 0;
}

.c5::-webkit-outer-spin-button,
.c5::-webkit-inner-spin-button {
  margin: 0;
  -webkit-appearance: none;
}

.c5[type="number"] {
  -moz-appearance: textfield;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  margin-bottom: 0;
}

.c2:after {
  content: '';
  display: table;
  clear: both;
}

.c11 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  margin-bottom: 5px;
}

.c8 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  color: #0b0c0c;
  margin-top: 0;
  margin-bottom: 15px;
  padding-left: 20px;
}

.c7 .c7 {
  margin-top: 10px;
}

.c8 > .c10 {
  margin-bottom: 5px;
}

.c21 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.c21:focus {
  outline: 3px solid #ffdd00;
  outline-offset: 0;
  background-color: #ffdd00;
}

.c21:link {
  color: #1d70b8;
}

.c21:visited {
  color: #4c2c92;
}

.c21:hover {
  color: #2b8cc4;
}

.c21:active {
  color: #2b8cc4;
}

.c21:focus {
  color: #0b0c0c;
}

.c1 {
  margin-bottom: 5px;
}

.c15 {
  cursor: pointer;
}

.c13 {
  background-color: white;
  border: 1px #D1D2D3 solid;
  border-radius: 5px;
  padding: 0.3125em 0.3125em;
  max-width: 100%;
  margin: 0.3125em 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c16 {
  word-wrap: break-word;
  padding-left: 0.5em;
  font-size: 16px;
  margin-bottom: 0;
}

.c14 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c18 {
  padding-right: 5px;
}

.c17 {
  padding: 5px 10px 5px 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: 300;
}

.c19 {
  color: #505A5F;
  padding-top: 1px;
}

.c9 {
  padding: 0px;
  margin: 0px;
  overflow: hidden;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c6 {
  display: inline-block;
  margin-top: 10px;
  font-family: Arial;
  margin-bottom: 5px;
  font-weight: 400;
  font-size: 19px;
  line-height: 1.415;
}

.c20 {
  padding-top: 10px;
  padding-bottom: 10px;
  font-size: 19px;
}

.c0 {
  margin: 0px;
  background-color: #FFFFFF;
}

@media print {
  .c4 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c4 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c3 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c3 {
    color: #000;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    margin-bottom: 0;
  }
}

@media print {
  .c5 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c5 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media only screen and (min-width:641px) {
  .c5 {
    margin-bottom: 0;
  }
}

@media only screen and (min-width:641px) {
  .c2 {
    margin-bottom: 0;
  }
}

@media print {
  .c11 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c11 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c8 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c8 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c8 {
    color: #000;
  }
}

@media only screen and (min-width:641px) {
  .c8 {
    margin-bottom: 20px;
  }
}

@media print {
  .c21 {
    font-family: sans-serif;
  }
}

@media print {
  .c21[href^="/"]::after,
  .c21[href^="http://"]::after,
  .c21[href^="https://"]::after {
    content: " (" attr(href) ")";
    font-size: 90%;
    word-wrap: break-word;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        class="c2"
        data-testid="search-form-input-area"
      >
        <span
          class="c3"
        >
          Search for an area
        </span>
        <span
          class="c4"
        >
          <div
            style="color: rgb(80, 90, 95);"
          >
            For example district, county, region, NHS organisation or GP practice or code
          </div>
        </span>
        <input
          class="c5"
          disabled=""
          value=""
        />
      </label>
    </div>
    <div>
      <span
        class="c6"
      >
        Selected areas (1)
      </span>
      <ul
        class="c7 c8 c9"
      >
         
        <li
          class="c10 c11 c12"
        >
          <div
            class="c13"
            data-testid="pill-container"
          >
            <div
              class="c14"
              data-testid="remove-icon-div"
            >
              <svg
                class="c15"
                data-testid="x-icon"
                fill="none"
                height="12"
                stroke="#000000"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="3"
                viewBox="0 0 25 25"
                width="12"
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="18"
                  x2="6"
                  y1="6"
                  y2="18"
                />
                <line
                  x1="6"
                  x2="18"
                  y1="6"
                  y2="18"
                />
              </svg>
            </div>
            <div
              class="c16"
              data-testid="filter-name"
            >
              <div
                class="c17"
              >
                <span
                  class="c18"
                >
                  Some area name
                </span>
                <span
                  class="c19"
                >
                  Type
                </span>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="c20"
    >
      <a
        class="c21"
        data-testid="search-form-link-filter-area"
        href="#"
      >
        Open a filter to add or change areas
      </a>
    </div>
  </div>
</DocumentFragment>
`;
